sigstore/oidc.py

```diff
@@ -56,6 +58,91 @@ class _OpenIDConfiguration(BaseModel):
     token_endpoint: StrictStr
 
 
+class IdentityToken:
+    """
+    An OIDC "identity", corresponding to an underlying OIDC token with
+    a sensible subject, issuer, and audience for Sigstore purposes.
+    """
+
+    def __init__(self, raw_token: str) -> None:
+        """
+        Create a new `IdentityToken` from the given OIDC token.
+        """
+
+        self._raw_token = raw_token
+
+        # NOTE: The lack of verification here is intentional, and is part of
+        # Sigstore's verification model: clients like sigstore-python are
+        # responsible only for forwarding the OIDC identity to Fulcio for
+        # certificate binding and issuance.
+        try:
+            identity_jwt = jwt.decode(
+                self._raw_token, options={"verify_signature": False}
+            )
+        except jwt.InvalidTokenError as exc:
+            raise IdentityError("invalid identity token") from exc
+
+        self._issuer: str = identity_jwt.get("iss")  # type: ignore
+        if self._issuer is None:
+            raise IdentityError("Identity token missing the required `iss` claim")
+
+        aud = identity_jwt.get("aud")
+        if aud is None:
+            raise IdentityError("Identity token missing the required `aud` claim")
+        if aud != _DEFAULT_AUDIENCE:
+            raise IdentityError(
+                f"Audience should be {_DEFAULT_AUDIENCE!r}, not {aud!r}"
+            )
+
+        # When verifying the private key possession proof, Fulcio uses
+        # different claims depending on the token's issuer.
+        # We currently special-case a handful of these, and fall back
+        # on signing the "sub" claim otherwise.
+        identity_claim = _KNOWN_OIDC_ISSUERS.get(self.issuer)
+        if identity_claim is not None:
+            if identity_claim not in identity_jwt:
+                raise IdentityError(
+                    f"Identity token missing the required {identity_claim!r} claim"
+                )
+
+            self._identity = str(identity_jwt.get(identity_claim))
+        else:
+            try:
+                self._identity = str(identity_jwt["sub"])
+            except KeyError:
+                raise IdentityError("Identity token missing the required 'sub' claim")
+
+    @property
+    def identity(self) -> str:
+        """
+        Returns this `IdentityToken`'s underlying subject.
+        """
+        return self._identity
+
+    @property
+    def issuer(self) -> str:
+        """
+        Returns a URL identifying this `IdentityToken`'s issuer.
+        """
+        return self._issuer
```

Comment:
I believe this returns the OIDC "issuer" which in a typical sigstore interactive login case is `https://oauth2.sigstore.dev/auth`. I think I am interested in the "federated connector" (or whatever the correct term is): the "issuer" that ends up written into the signing certificate itself, e.g. `https://github.com/login/oauth`

Reply:
I'll double-check on this!

Reply:
Yep, confirmed.

@jku would a `federated_connector_id: str | None` work for you here? I assume we'll need to accommodate the `None` case for identity tokens that don't go through a second hoop like the Sigstore/Sigstage OAuth connector.

Reply:
So this means there are situations where the application would have to somehow choose when too look at `federated_connector_id` and when to look at `issuer`? It feels like this still requires the application to make decisions that require knowing sigstore implementation details, and that the sigstore library would be a better place for that logic: If it's possible I would rather have a `certificate_issuer` (or something that always contains the issuer that will be inserted into the signing certificate).

Reply:
> So this means there are situations where the application would have to somehow choose when too look at `federated_connector_id` and when to look at `issuer`? 

Yes, I think so.

> It feels like this still requires the application to make decisions that require knowing sigstore implementation details, and that the sigstore library would be a better place for that logic: If it's possible I would rather have a `certificate_issuer` (or something that always contains the issuer that will be inserted into the signing certificate).

`certificate_issuer` sounds a little funny to me here (I could be underthinking it!) since the identity token at this point is conceptually untied to any certificate issuance -- Fulcio *itself* is the issuer and will issue a certificate with `federated_connector_id` as an *extension* value, but only if it happens to accept this particular OIDC token (which we can't guarantee a priori).

I'm not sure if there's a clean way for `sigstore-python` to handle this for users, unless we get additional guarantees from Fulcio about whether `federated_connector_id` is _always_ the value that ends up becoming the underlying issuer extension value. If they can guarantee that, then IMO an attribute like `federated_issuer` or similar would make sense here.

CC @haydentherapper and @asraa for that last point 🙂

Reply:
(Separately, I think a user could still use this without knowing too much about Sigstore's internals: if a user would like to completely forbid an IdP at `hxxps://example.com`, then *neither* `issuer` nor `federated_connector_id` should ever match that value.)

Reply:
Very interesting points!

The federated connector ID is an implementation detail of Sigstore and one that I wouldn’t expect clients would need to be aware of. I would assume clients simply pass through the identity token they’re given and the receiving party (fulcio) will handle the claims how it chooses. 

If a client wanted to verify the identity token before use, then the token no longer is just a blob and you’d need to know the expected claims per issuer. This is hard though because there’s so many different permutations - a subject may be in sub, or email (or for CI, so many options!). Issuer is in iss, but should the client know that dex is a federated provider and therefore you need to also look into the federated claim? 

 My overall comment is “it depends” to whether the client should know about federation or not. The simplest view is that an ID token is just a blob that gets returned from the OIDC dance and passed to fulcio. If you do want to verify a token before use, maybe you push this to a policy later, and let users specify which claims they want to verify? Otherwise you could have a public good infrastructure default policy that’s aware of federated issuer, but there’s still the question of where identity comes from.

Related, how are you picking which subject to sign for the challenge? Or do you just use CSRs, which don’t require a particular subject signed over?

Reply:
Thanks for the detailed response @haydentherapper!

Yeah, this is a hard position for the client to be in: @jku has presented a reasonable use case in which a user might want to sanity-check an OIDC identity before sending it along to Fulcio, but I don't think there's a straightforward way for us to consistently guarantee the presence of the fields that the user might expect (either because of Dex, or an unusual but valid claim set, etc.).

I _think_ the best we can do is exposing the `federated_issuer` when we happen to have it, but it's possibly that that's also a Dex implementation detail and I'm not realizing that. 


> Related, how are you picking which subject to sign for the challenge? Or do you just use CSRs, which don’t require a particular subject signed over?

Yep, we punt on this by using CSRs :slightly_smiling_face:

Reply:
I’d lean toward it being an implementation detail. To my knowledge there’s no IETF standard for this concept, so we are exposing a detail that’s only present for the public instance (or someone running their own dex instance). 
it’s simple enough to add though, so I’m supportive of that, and if we see no usage of it, we can always clean it up later.

Reply:
Opened https://github.com/sigstore/fulcio/issues/1145 for this.

Reply:
Will respond there, thanks!